### PR TITLE
Add shared ViewModel and card with Review Progress

### PR DESCRIPTION
**Pull request by Vladuken**
Смотреть после этого пул реквеста:
https://github.com/Android-Academy-Global/advanced-2021-architecture-1/pull/13

### CODE REVIEW

File: app/src/main/java/ru/gaket/themoviedb/core/data/HeapItemStore.kt [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+        synchronizedWrite(lock) {
+            val current = this._item ?: return
+            updated = updateAction.invoke(current)
+            this._item = updated
+        }
+
+        _itemChanges.emit(updated)
+    }
+
+    override suspend fun reset() {
```
Comment (rHlushan) [inline app/src/main/java/ru/gaket/themoviedb/core/data/HeapItemStore.kt:50]: just create
fun setValue(item: T?)
and reuse it for
override suspend fun setItem(item: T)
override suspend fun reset()

File: app/src/main/java/ru/gaket/themoviedb/core/data/HeapItemStore.kt [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+import ru.gaket.themoviedb.domain.review.store.ItemStore
+import ru.gaket.themoviedb.util.synchronizedRead
+import ru.gaket.themoviedb.util.synchronizedWrite
+import java.util.concurrent.locks.ReentrantReadWriteLock
+
+private const val BUFFER_CAPACITY = 16
+
+class HeapItemStore<T : Any>(replay: Int) : ItemStore<T> {
+
+    private var _item: T? = null
```
Comment (rHlushan) [inline app/src/main/java/ru/gaket/themoviedb/core/data/HeapItemStore.kt:14]: why not to use just StateFlow?
I know about their diff

but generally is this approach with override suspend fun SharedFlow as changes listener really better in current case?

File: app/src/main/java/ru/gaket/themoviedb/core/data/HeapItemStore.kt [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    private val _itemChanges = MutableSharedFlow<T?>(
+        replay = replay,
+        extraBufferCapacity = BUFFER_CAPACITY,
+    )
+    private val lock = ReentrantReadWriteLock()
```
Comment (rHlushan) [inline app/src/main/java/ru/gaket/themoviedb/core/data/HeapItemStore.kt:19]: using just atomic is easier here

File: app/src/main/java/ru/gaket/themoviedb/data/movies/MoviesRepository.kt [hunk 4]
```diff
@@ -97,6 +94,29 @@ class MoviesRepositoryImpl @Inject constructor(
         searchMovies(query, page)
             .mapSuccess { movies -> fillSearchMoviesWithMyReviews(movies) }
 
+    private suspend fun searchMovies(query: String, page: Int): OperationResult<List<SearchMovie>, Throwable> =
```
Comment (rHlushan) [inline app/src/main/java/ru/gaket/themoviedb/data/movies/MoviesRepository.kt:97]: why did u moved private functions above public?

File: app/src/main/java/ru/gaket/themoviedb/data/review/repository/ReviewRepositoryImpl.kt [hunk 5]
```diff
@@ -1,59 +1,58 @@
...
-            AddReviewRequest(
-                movieId = reviewFormStore.item.movieId ?: error("movieId is not provided"),
-                liked = reviewFormStore.item.whatLiked ?: error("whatLiked is not provided"),
-                disliked = reviewFormStore.item.whatDidNotLike ?: error("whatDidNotLike is not provided"),
-                rating = reviewFormStore.item.rating ?: error("rating is not provided")
-            )
-        }.doOnError {
-            Timber.e(it)
-        }
+    override fun buildReview(): AddReviewRequest {
```
Comment (rHlushan) [inline app/src/main/java/ru/gaket/themoviedb/data/review/repository/ReviewRepositoryImpl.kt:44]: why this was changed?

File: app/src/main/java/ru/gaket/themoviedb/presentation/review/ReviewFragment.kt [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+    override fun onAttach(context: Context) {
+        super.onAttach(context)
+        requireActivity().onBackPressedDispatcher.addCallback(this, backPressedCallback)
+    }
+
+    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+        super.onViewCreated(view, savedInstanceState)
+        viewModel.reviewState.observe(viewLifecycleOwner) {
+            val containerId = binding.containerReview.id
+            when (it) {
```
Comment (rHlushan) [inline app/src/main/java/ru/gaket/themoviedb/presentation/review/ReviewFragment.kt:56]: please move this to private fun

File: app/src/main/java/ru/gaket/themoviedb/presentation/review/ReviewFragment.kt [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+
+    override fun onAttach(context: Context) {
+        super.onAttach(context)
+        requireActivity().onBackPressedDispatcher.addCallback(this, backPressedCallback)
+    }
+
+    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+        super.onViewCreated(view, savedInstanceState)
+        viewModel.reviewState.observe(viewLifecycleOwner) {
+            val containerId = binding.containerReview.id
```
Comment (rHlushan) [inline app/src/main/java/ru/gaket/themoviedb/presentation/review/ReviewFragment.kt:55]: add @IdRes

File: app/src/main/java/ru/gaket/themoviedb/presentation/review/ReviewFragment.kt [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+            }
+        }
+
+        viewModel.currentReview.observe(viewLifecycleOwner) { movieToReview ->
+            binding.cvReviewSummary.isVisible = movieToReview != null
+            movieToReview?.let { (movie, review) -> binding.bind(movie, review) }
+        }
+    }
+
+    private fun FragmentReviewBinding.bind(movie: Movie, review: ReviewFormModel) {
```
Comment (rHlushan) [inline app/src/main/java/ru/gaket/themoviedb/presentation/review/ReviewFragment.kt:76]: please avoid nested extensions:
u can move this out of fragment class

File: app/src/main/java/ru/gaket/themoviedb/presentation/review/whatliked/WhatLikeFragment.kt [hunk 9]
```diff
@@ -40,17 +35,12 @@ class WhatLikeFragment : Fragment(R.layout.fragment_review_text) {
...
         }
-    }
 
-    companion object {
-
-        fun newInstance(movieId: MovieId): WhatLikeFragment {
-            return WhatLikeFragment().apply {
-                arguments = bundleOf(WhatLikeViewModel.ARG_MOVIE_ID to movieId)
-            }
+        viewModel.initialValue.observe(viewLifecycleOwner) { initialValue ->
```
Comment (rHlushan) [inline app/src/main/java/ru/gaket/themoviedb/presentation/review/whatliked/WhatLikeFragment.kt:42]: maybe use use function reference?

File: app/src/main/java/ru/gaket/themoviedb/presentation/review/whatnotliked/WhatNotLikeFragment.kt [hunk 10]
```diff
@@ -38,8 +35,12 @@ class WhatNotLikeFragment : Fragment(R.layout.fragment_review_text) {
         viewModel.events.observe(viewLifecycleOwner) { reviewErrorField ->
             when (reviewErrorField) {
                 EMPTY_FIELD -> requireView().showSnackbar(R.string.review_error_should_not_be_empty)
-                SUCCESS -> navigator.navigateTo(RatingScreen())
+                SUCCESS -> sharedViewModel.nextState()
             }
         }
+
+        viewModel.initialValue.observe(viewLifecycleOwner) { initialValue ->
```
Comment (rHlushan) [inline app/src/main/java/ru/gaket/themoviedb/presentation/review/whatnotliked/WhatNotLikeFragment.kt:42]: function ref

File: app/src/main/java/ru/gaket/themoviedb/presentation/review/rating/RatingFragment.kt [hunk 11]
```diff
@@ -18,19 +17,17 @@ import ru.gaket.themoviedb.presentation.review.rating.RatingViewModel.ReviewEven
...
 class RatingFragment : Fragment(R.layout.fragment_review_rating) {
 
-    @Inject
-    lateinit var navigator: Navigator
-
     private val binding: FragmentReviewRatingBinding by viewBinding(FragmentReviewRatingBinding::bind)
 
     private val viewModel: RatingViewModel by viewModels()
 
+    private val sharedViewModel: ReviewViewModel by viewModels({ requireParentFragment() })
```
Comment (rHlushan) [inline app/src/main/java/ru/gaket/themoviedb/presentation/review/rating/RatingFragment.kt:29]: both brackets are required?

File: app/src/main/res/layout/fragment_review.xml [hunk 12]
```diff
@@ -0,0 +1,35 @@
+<?xml version="1.0" encoding="utf-8"?>
+<androidx.constraintlayout.widget.ConstraintLayout xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:app="http://schemas.android.com/apk/res-auto"
+    android:layout_width="match_parent"
+    android:layout_height="match_parent">
+
+    <androidx.cardview.widget.CardView
```
Comment (rHlushan) [inline app/src/main/res/layout/fragment_review.xml:7]: maybe use single approach everywhere with snake case of camel case for ids?

**Comment by Gaket** [comment]
There were some nice comments but it's time to merge the project

**Pull request merged**